pkg/archive/archive.go

```diff
@@ -231,13 +231,51 @@ func (r *readCloserWrapper) Close() error {
 
 		return nil
 	}
-	return r.closer()
+	if r.closer != nil {
+		return r.closer()
+	}
+	return nil
+}
+
+var (
+	bufioReader32KPool = &sync.Pool{
+		New: func() interface{} { return bufio.NewReaderSize(nil, 32*1024) },
+	}
+)
+
+type bufferedReader struct {
+	buf *bufio.Reader
+}
+
+func newBufferedReader(r io.Reader) *bufferedReader {
+	buf := bufioReader32KPool.Get().(*bufio.Reader)
+	buf.Reset(r)
+	return &bufferedReader{buf}
+}
+
+func (r *bufferedReader) Read(p []byte) (n int, err error) {
+	if r.buf == nil {
+		return 0, io.EOF
+	}
+	n, err = r.buf.Read(p)
+	if err == io.EOF {
```

Comment:
Won't this leak the buffer if read fails with other errors? Couldn't this just be `err != nil`?

Reply:
Guess this should have a closer to return the buffer

Reply:
In that case it will just get GC'ed. It is pointless to return to the pool on other errors since the higher level operation will end. Am I missing something else which will hold other a reference?

This implementation is intentionally the same as containerd's. We can probably unify the two, the only difference is we don't (intentionally) support xz.